packs/orion/actions/lib/actions.py

```diff
@@ -0,0 +1,212 @@
+# Licensed to the StackStorm, Inc ('StackStorm') under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import time
+
+from st2actions.runners.pythonrunner import Action
+from orionsdk import SwisClient
+
+
+class OrionBaseAction(Action):
+    def __init__(self, config):
+        super(OrionBaseAction, self).__init__(config)
+
+        self.client = None
+
+        if "orion" not in self.config:
+            raise ValueError("Orion host details not in the config.yaml")
+
+    def connect(self, platform):
+        """
+        Connect to an Orion platform from the packs config.yaml.
+        """
+        try:
+            self.client = SwisClient(
+                self.config['orion'][platform]['host'],
+                self.config['orion'][platform]['user'],
+                self.config['orion'][platform]['password'])
+        except KeyError:
+            raise ValueError("Orion host details not in the config.yaml")
+
+    def query(self, swql, **kargs):
+        """
+        Run SWQL against the Orion Platform.
+        """
+        return self.client.query(swql, **kargs)
+
+    def invoke(self, entity, verb, *args):
+        """
+        Run an Invoke against the Orion Platform.
+        """
+        return self.client.invoke(entity, verb, *args)
+
+    def create(self, entity, **kargs):
+        """
+        Run an Create against the Orion Platform.
+        """
+        return self.client.create(entity, **kargs)
+
+    def node_exists(self, caption, ip_address):
+        """
+        Check if an Node exists (caption and or ip) on the Orion platform.
+
+        Returns: True or False.
+        """
+        swql = """SELECT NodeID, IPAddress FROM Orion.Nodes
+                  WHERE Caption=@caption"""
+        kargs = {'caption': caption}
+        caption_data = self.query(swql, **kargs)
+
+        swql = """SELECT NodeID, IPAddress FROM Orion.Nodes
+                  WHERE IPAddress=@ip_address"""
+        kargs = {'ip_address': ip_address}
+        ip_data = self.query(swql, **kargs)
+
+        if len(caption_data['results']) >= 1 or len(ip_data['results']) >= 1:
```

Comment:
Would it make sense to re-arrange this a bit and return immediately if first query returns any result?

This way we could reduce total numbers of queries from 2 to 1 in case the first query already matches.

In any case, that's not a big deal, probably just a pre-mature optimization :)